src/mame/sony/news_020_mmu.cpp

```diff
@@ -0,0 +1,298 @@
+{
+	m_bus_error.resolve();
+
+	save_item(NAME(m_enabled));
+	save_item(NAME(m_romdis));
+
+	m_mmu_user_ram = std::make_unique<u32[]>(MMU_ENTRY_COUNT);
+	save_pointer(NAME(m_mmu_user_ram), MMU_ENTRY_COUNT);
+	m_mmu_system_ram = std::make_unique<u32[]>(MMU_ENTRY_COUNT);
+	save_pointer(NAME(m_mmu_system_ram), MMU_ENTRY_COUNT);
+
+	m_mmu_user_tag_ram = std::make_unique<u32[]>(MMU_ENTRY_COUNT);
+	save_pointer(NAME(m_mmu_user_tag_ram), MMU_ENTRY_COUNT);
+	m_mmu_system_tag_ram = std::make_unique<u32[]>(MMU_ENTRY_COUNT);
+	save_pointer(NAME(m_mmu_system_tag_ram), MMU_ENTRY_COUNT);
+}
+
+device_memory_interface::space_config_vector news_020_mmu_device::memory_space_config() const
+{
+	return space_config_vector{
+		std::make_pair(AS_PROGRAM, &m_hyperbus_config)};
+}
+
+uint32_t news_020_mmu_device::mmu_entry_r(offs_t offset, uint32_t mem_mask)
+{
+	fatalerror("news_020_mmu: tried to read from mapping RAM!");
+	return 0;
+}
+
+void news_020_mmu_device::mmu_entry_w(offs_t offset, uint32_t data, uint32_t mem_mask)
+{
+	const bool system_map = offset & 0x4000000; // user or system mapping
+	const uint32_t tag = (offset & ~0x4000000);
+
+	// Actually update map
+	offset = offset % MMU_ENTRY_COUNT;
+	if (system_map)
+	{
+		COMBINE_DATA(&m_mmu_system_ram[offset]);
+		m_mmu_system_tag_ram[offset] = (m_mmu_system_tag_ram[offset] & ~mem_mask) | (tag & mem_mask);
+	}
+	else
+	{
+		COMBINE_DATA(&m_mmu_user_ram[offset]);
+		m_mmu_user_tag_ram[offset] = (m_mmu_user_tag_ram[offset] & ~mem_mask) | (tag & mem_mask);
+	}
+
+#if (VERBOSE & LOG_ENTRY) > 0
+	news_020_pte modified_entry(system_map ? m_mmu_system_ram[offset] : m_mmu_user_ram[offset]);
+	LOGMASKED(LOG_ENTRY, "(%s) MMU %s entry write (0x%08x, 0x%08x, 0x%08x) -> entry 0x%08x: tag = 0x%08x, pfnum = 0x%08x (valid = %d, modified = %d, fill_on_demand = %d, kernel write = %d, kernel read = %d, user write = %d, user read = %d)\n",
+			  machine().describe_context(),
+			  system_map ? "system" : "user",
+			  offset,
+			  data,
+			  mem_mask,
+			  offset % MMU_ENTRY_COUNT,
+			  tag,
+			  modified_entry.pfnum(),
+			  modified_entry.valid(),
+			  modified_entry.modified(),
+			  modified_entry.fill_on_demand(),
+			  modified_entry.kernel_writable(),
+			  modified_entry.kernel_readable(),
+			  modified_entry.user_writable(),
+			  modified_entry.user_readable());
+#endif
+}
+
+uint32_t news_020_mmu_device::hyperbus_r(offs_t offset, uint32_t mem_mask, bool is_supervisor)
+{
+	uint32_t result = 0;
+	if (!m_enabled || (is_supervisor && (offset >= (0xc0000000 >> 2)))) // TODO: is the physical address check here (0xc...) correct?
+	{
+		// TODO: What is the actual physical address width? The 0x1 part might also be an MMU control bit
+		offset = offset & (0x1fffffff >> 2);
+		if (!m_romdis && offset < ((0xffff - 0x100) >> 2))
+		{
+			result = this->space(0).read_dword((0x03000000 + 0x100) + (offset << 2), mem_mask);
+		}
+		else
+		{
+			result = this->space(0).read_dword(offset << 2, mem_mask);
+		}
```

Comment:
Can you at least use a `memory_access::specific<...>` to avoid going through the slowest path for accessing the address space?